File: android/src/main/kotlin/com/clickio/clickio_consent_sdk/ClickioConsentSdkPlugin.kt
package com.clickio.clickio_consent_sdk

import android.content.Context
import android.view.View
import android.view.ViewGroup
import androidx.fragment.app.FragmentActivity
import com.clickio.clickioconsentsdk.ClickioConsentSDK
import com.clickio.clickioconsentsdk.ClickioConsentSDK.Config
import com.clickio.clickioconsentsdk.WebViewConfig
import com.clickio.clickioconsentsdk.WebViewGravity
import com.clickio.clickioconsentsdk.LogsMode
import com.clickio.clickioconsentsdk.ExportData
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.activity.ActivityAware
import io.flutter.embedding.engine.plugins.activity.ActivityPluginBinding
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import io.flutter.plugin.platform.PlatformViewFactory
import io.flutter.plugin.common.StandardMessageCodec
import io.flutter.plugin.common.MethodChannel.Result
import io.flutter.plugin.common.MethodChannel.MethodCallHandler
import android.util.Log
import java.util.Locale

class ClickioWebViewFactory(
    private val fragmentActivity: FragmentActivity
) : PlatformViewFactory(StandardMessageCodec.INSTANCE) {

    override fun create(context: Context, viewId: Int, args: Any?): PlatformView {
        val params = args as? Map<String, Any?> ?: emptyMap()

        val url = params["url"] as? String ?: ""
        val backgroundColor = (params["backgroundColor"] as? Number)?.toInt() ?: 0
        val height = params["height"] as? Int ?: -1
        val width = params["width"] as? Int ?: -1        
        val gravity: WebViewGravity = (params["gravity"] as? String)?.let {
            runCatching { WebViewGravity.valueOf(it.uppercase(Locale.ROOT)) }.getOrNull()
        } ?: WebViewGravity.CENTER

        val webViewConfig = WebViewConfig(
            backgroundColor = backgroundColor,
            height = height,
            width = width,
            gravity = gravity
        )

        val webView = ClickioConsentSDK.getInstance().webViewLoadUrl(
            fragmentActivity,
            url,
            webViewConfig
        )

        return object : PlatformView {
            override fun getView(): View = webView
            override fun dispose() {}
        }
    }
}

class ClickioConsentSdkPlugin : FlutterPlugin, MethodCallHandler, ActivityAware {
    private lateinit var channel: MethodChannel
    private var fragmentActivity: FragmentActivity? = null
    private var pluginBinding: FlutterPlugin.FlutterPluginBinding? = null

    override fun onAttachedToEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        pluginBinding = binding
        channel = MethodChannel(binding.binaryMessenger, "clickio_consent_sdk")
        channel.setMethodCallHandler(this)
    }

    override fun onDetachedFromEngine(binding: FlutterPlugin.FlutterPluginBinding) {
        channel.setMethodCallHandler(null)
        pluginBinding = null
    }

    override fun onAttachedToActivity(binding: ActivityPluginBinding) {
        val act = binding.activity
        if (act is FragmentActivity) {
            fragmentActivity = act

            // Register the PlatformViewFactory now that we have the FragmentActivity
            pluginBinding?.platformViewRegistry?.registerViewFactory(
                "clickio_webview",
                ClickioWebViewFactory(act)
            )
        } else {
            throw IllegalStateException("Activity must be a FragmentActivity")
        }
    }

    override fun onDetachedFromActivityForConfigChanges() {
        fragmentActivity = null
    }

    override fun onReattachedToActivityForConfigChanges(binding: ActivityPluginBinding) {
        onAttachedToActivity(binding)
    }

    override fun onDetachedFromActivity() {
        fragmentActivity = null
    }

    override fun onMethodCall(call: MethodCall, result: Result) {
        when (call.method) {
            "initialize" -> initialize(call, result)
            "setLogsMode" -> setLogsMode(call, result)
            "openDialog" -> openDialog(call, result)
            "getConsentScope" -> getConsentScope(result)
            "getConsentState" -> getConsentState(result)
            "getConsentForPurpose" -> getConsentForPurpose(call, result)
            "getConsentForVendor" -> getConsentForVendor(call, result)
            "getTCString" -> getTCString(result)
            "getACString" -> getACString(result)
            "getGPPString" -> getGPPString(result)
            "getConsentedTCFVendors" -> getConsentedTCFVendors(result)
            "getConsentedTCFLiVendors" -> getConsentedTCFLiVendors(result)
            "getConsentedTCFPurposes" -> getConsentedTCFPurposes(result)
            "getConsentedTCFLiPurposes" -> getConsentedTCFLiPurposes(result)
            "getConsentedGoogleVendors" -> getConsentedGoogleVendors(result)
            "getConsentedOtherVendors" -> getConsentedOtherVendors(result)
            "getConsentedOtherLiVendors" -> getConsentedOtherLiVendors(result)
            "getConsentedNonTcfPurposes" -> getConsentedNonTcfPurposes(result)
            "getGoogleConsentMode" -> getGoogleConsentMode(result)
            else -> result.notImplemented()
        }
    }

    private fun initialize(call: MethodCall, result: Result) {
        val act = fragmentActivity ?: return result.error("NO_ACTIVITY", "FragmentActivity is null", null)
        val siteId: String = call.argument<String>("siteId")
            ?: return result.error("INVALID_ARGUMENT", "siteId is required", null)
        val language = call.argument<String>("language") ?: "en"
        val config = Config(siteId, language)

        val clickioConsentSdk = ClickioConsentSDK.getInstance()
        clickioConsentSdk.initialize(act, config)

        clickioConsentSdk.onReady {
            channel.invokeMethod("onReady", null)
        }

        clickioConsentSdk.onConsentUpdated {
            channel.invokeMethod("onConsentUpdate", null)
        }

        result.success(null)
    }

    private fun setLogsMode(call: MethodCall, result: Result) {
        val mode = when (call.argument<String>("mode")) {
            "verbose" -> LogsMode.VERBOSE
            else -> LogsMode.DISABLED
        }
        ClickioConsentSDK.getInstance().setLogsMode(mode)
        result.success(null)
    }

    private fun openDialog(call: MethodCall, result: Result) {
        val act = fragmentActivity ?: return result.error("NO_ACTIVITY", "FragmentActivity is null", null)
        val mode = when (call.argument<String>("mode")) {
            "resurface" -> ClickioConsentSDK.DialogMode.RESURFACE
            else -> ClickioConsentSDK.DialogMode.DEFAULT
        }
        ClickioConsentSDK.getInstance().openDialog(act, mode)
        result.success(null)
    }

    private fun getConsentScope(result: Result) = result.success(ClickioConsentSDK.getInstance().checkConsentScope().toString())
   
    private fun getConsentState(result: Result) = result.success(ClickioConsentSDK.getInstance().checkConsentState().toString())
   
    private fun getConsentForPurpose(call: MethodCall, result: Result) {
        val id = call.argument<Int>("id") ?: 1
        result.success(ClickioConsentSDK.getInstance().checkConsentForPurpose(id).toString())
    }
   
    private fun getConsentForVendor(call: MethodCall, result: Result) {
        val id = call.argument<Int>("id") ?: 9
        result.success(ClickioConsentSDK.getInstance().checkConsentForVendor(id).toString())
    }
   
    private fun getTCString(result: Result) { result.success(ExportData(fragmentActivity!!).getTCString()) }
   
    private fun getACString(result: Result) { result.success(ExportData(fragmentActivity!!).getACString()) }
   
    private fun getGPPString(result: Result) { result.success(ExportData(fragmentActivity!!).getGPPString().toString()) }
   
    private fun getGoogleConsentMode(result: Result) { result.success(ExportData(fragmentActivity!!).getGoogleConsentMode().toString()) }
   
    private fun getConsentedTCFVendors(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedTCFVendors().toString()) }
   
    private fun getConsentedTCFLiVendors(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedTCFLiVendors().toString()) }
   
    private fun getConsentedTCFPurposes(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedTCFPurposes().toString()) }
   
    private fun getConsentedTCFLiPurposes(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedTCFLiPurposes().toString()) }
   
    private fun getConsentedGoogleVendors(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedGoogleVendors().toString()) }
   
    private fun getConsentedOtherVendors(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedOtherVendors().toString()) }
   
    private fun getConsentedOtherLiVendors(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedOtherLiVendors().toString()) }
   
    private fun getConsentedNonTcfPurposes(result: Result) { result.success(ExportData(fragmentActivity!!).getConsentedNonTcfPurposes().toString()) }
}

File: ios/Classes/ClickioConsentSdkPlugin.swift
import ClickioConsentSDKManager
import Flutter
import UIKit

class ClickioWebViewFactory: NSObject, FlutterPlatformViewFactory {
  func create(
    withFrame frame: CGRect,
    viewIdentifier viewId: Int64,
    arguments args: Any?
  ) -> FlutterPlatformView {
    let params = args as? [String: Any] ?? [:]
      
    guard let urlString = params["url"] as? String, URL(string: urlString) != nil,
      !urlString.isEmpty
    else {
      print("ClickioWebViewFactory: invalid URL, returning empty view")

      return ClickioWebViewWrapper(webView: UIView())  
    }

    let backgroundColorValue = params["backgroundColor"] as? NSNumber ?? 0xFFFF_FFFF
    let backgroundColor = UIColor(
      red: CGFloat((backgroundColorValue.intValue >> 16) & 0xFF) / 255.0,
      green: CGFloat((backgroundColorValue.intValue >> 8) & 0xFF) / 255.0,
      blue: CGFloat(backgroundColorValue.intValue & 0xFF) / 255.0,
      alpha: CGFloat((backgroundColorValue.intValue >> 24) & 0xFF) / 255.0
    )

    let height: CGFloat? = {
        if let heightValue = params["height"] as? Int {
            return CGFloat(heightValue)
        }

        return nil
    }()

    let width: CGFloat? = {
        if let widthValue = params["width"] as? Int {
            return CGFloat(widthValue)
        }
        
        return nil
    }()

    let gravity: WebViewGravity = {
    if let gravityString = params["gravity"] as? String {
        switch gravityString.lowercased() {
        case "top":
            return .top
        case "center":
            return .center
        case "bottom":
            return .bottom
        default:
            return .center
        }
    }
    return .center
    }()

    let webViewConfig = WebViewConfig(
      backgroundColor: backgroundColor,
      width: width,
      height: height,
      gravity: gravity
    )

    let webViewController = ClickioConsentSDK.shared.webViewLoadUrl(
      urlString: urlString,
      config: webViewConfig
    )

    return ClickioWebViewWrapper(webView: webViewController.view)
  }
    
    func createArgsCodec() -> FlutterMessageCodec & NSObjectProtocol {
        return FlutterStandardMessageCodec.sharedInstance()
    }
}

class ClickioWebViewWrapper: NSObject, FlutterPlatformView {
  private let webView: UIView

  init(webView: UIView) {
    self.webView = webView
  }

  func view() -> UIView {
    return webView
  }
}

public class ClickioConsentSdkPlugin: NSObject, FlutterPlugin {
  private var channel: FlutterMethodChannel?

  public static func register(with registrar: FlutterPluginRegistrar) {
    let methodChannel = FlutterMethodChannel(
      name: "clickio_consent_sdk",
      binaryMessenger: registrar.messenger()
    )

    let factory = ClickioWebViewFactory()
    registrar.register(factory, withId: "clickio_webview")

    let instance = ClickioConsentSdkPlugin()
    registrar.addMethodCallDelegate(instance, channel: methodChannel)

    ClickioConsentSDK.shared.onReady {
      methodChannel.invokeMethod("onReady", arguments: nil)
    }

    ClickioConsentSDK.shared.onConsentUpdated {
      methodChannel.invokeMethod("onConsentUpdate", arguments: nil)
    }
  }

  @MainActor
  public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    switch call.method {
    case "initialize":
      self.initializeConsentSDK(call: call, result: result)
    case "setLogsMode":
      self.setLogsMode(call, result: result)
    case "openDialog":
      guard
        let controller = UIApplication.shared.windows.first(where: \.isKeyWindow)?
          .rootViewController
      else {
        result(
          FlutterError(
            code: "NO_VIEW_CONTROLLER",
            message: "Root view controller not found",
            details: nil))
        return
      }
      self.openDialog(call: call, viewController: controller, result: result)
    case "getConsentScope":
      self.getConsentScope(result: result)
    case "getConsentState":
      self.getConsentState(result: result)
    case "getConsentForPurpose":
      self.getConsentForPurpose(call: call, result: result)
    case "getConsentForVendor":
      self.getConsentForVendor(call: call, result: result)
    case "getTCString":
      self.getTCString(result: result)
    case "getACString":
      self.getACString(result: result)
    case "getGPPString":
      self.getGPPString(result: result)
    case "getConsentedTCFVendors":
      self.getConsentedTCFVendors(result: result)
    case "getConsentedTCFLiVendors":
      self.getConsentedTCFLiVendors(result: result)
    case "getConsentedTCFPurposes":
      self.getConsentedTCFPurposes(result: result)
    case "getConsentedTCFLiPurposes":
      self.getConsentedTCFLiPurposes(result: result)
    case "getConsentedGoogleVendors":
      self.getConsentedGoogleVendors(result: result)
    case "getConsentedOtherVendors":
      self.getConsentedOtherVendors(result: result)
    case "getConsentedOtherLiVendors":
      self.getConsentedOtherLiVendors(result: result)
    case "getConsentedNonTcfPurposes":
      self.getConsentedNonTcfPurposes(result: result)
    case "getGoogleConsentMode":
      self.getGoogleConsentMode(result: result)
    default:
      print("Unsupported method: \(call.method)")
      result(FlutterMethodNotImplemented)
    }
  }

  private func initializeConsentSDK(call: FlutterMethodCall, result: @escaping FlutterResult) {
    let args = call.arguments as? [String: Any]
    let siteId = args?["siteId"] as? String ?? ""
    let language = args?["language"] as? String ?? "en"

    if siteId.isEmpty {
      result(FlutterError(code: "MISSING_ARGUMENTS", message: "'siteId' is required", details: nil))
      return
    }

    let config = ClickioConsentSDK.Config(siteId: siteId, appLanguage: language)

    Task {
      do {
        await ClickioConsentSDK.shared.initialize(configuration: config)
        result("ClickioConsentSDK initialized successfully")
      }
    }
  }

  @MainActor
  private func setLogsMode(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let logMode = args["mode"] as? String
    else {
      return result(
        FlutterError(
          code: "INVALID_ARGUMENTS",
          message: "Missing or invalid arguments for setLogsMode",
          details: nil))
    }

    let mode: EventLogger.Mode

    switch logMode {
    case "verbose":
      mode = .verbose
    case "disabled":
      mode = .disabled
    default:
      return result(
        FlutterError(
          code: "INVALID_MODE",
          message: "Invalid logging mode: \(logMode)",
          details: nil))
    }

    ClickioConsentSDK.shared.setLogsMode(mode)

    result(nil)
  }

  @MainActor
  private func openDialog(
    call: FlutterMethodCall, viewController: UIViewController, result: @escaping FlutterResult
  ) {
    guard let args = call.arguments as? [String: Any],
      let dialogMode = args["mode"] as? String,
      let attNeeded = args["attNeeded"] as? Bool
    else {
      return result(
        FlutterError(code: "INVALID_ARGUMENTS", message: "Invalid dialog parameters", details: nil))
    }

    let mode: ClickioConsentSDK.DialogMode = dialogMode == "resurface" ? .resurface : .default

    DispatchQueue.main.async {
      do {
        ClickioConsentSDK.shared.openDialog(
          mode: mode,
          in: viewController,
          attNeeded: attNeeded
        )

        result("Consent Dialog Opened")
      }
    }
  }

  @MainActor
  private func getConsentScope(result: FlutterResult) {
    result(ClickioConsentSDK.shared.checkConsentScope())
  }

  @MainActor
  private func getConsentState(result: FlutterResult) {
    if let consentState = ClickioConsentSDK.shared.checkConsentState() {
      result(consentState.rawValue)
    } else {
      result(
        FlutterError(
          code: "CONSENT_STATE_ERROR",
          message: "Failed to retrieve consent state",
          details: nil
        ))
    }
  }

  @MainActor
  private func getConsentForPurpose(call: FlutterMethodCall, result: FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let purposeId = args["id"] as? Int
    else {
      result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing purposeId", details: nil))
      return
    }

    result(ClickioConsentSDK.shared.checkConsentForPurpose(purposeId: purposeId)?.description)
  }

  @MainActor
  private func getConsentForVendor(call: FlutterMethodCall, result: FlutterResult) {
    guard let args = call.arguments as? [String: Any],
      let vendorId = args["id"] as? Int
    else {
      result(FlutterError(code: "ARGUMENT_ERROR", message: "Missing vendorId", details: nil))
      return
    }

    result(ClickioConsentSDK.shared.checkConsentForVendor(vendorId: vendorId)?.description)
  }

  private func getTCString(result: FlutterResult) {
    result(ExportData().getTCString())
  }

  private func getACString(result: FlutterResult) {
    result(ExportData().getACString())
  }

  private func getGPPString(result: FlutterResult) {
    result(ExportData().getGPPString())
  }

  private func getConsentedTCFVendors(result: FlutterResult) {
    result(ExportData().getConsentedTCFVendors()?.description)
  }

  private func getConsentedTCFLiVendors(result: FlutterResult) {
    result(ExportData().getConsentedTCFLiVendors()?.description)
  }

  private func getConsentedTCFPurposes(result: FlutterResult) {
    result(ExportData().getConsentedTCFPurposes()?.description)
  }

  private func getConsentedTCFLiPurposes(result: FlutterResult) {
    result(ExportData().getConsentedTCFLiPurposes()?.description)
  }

  private func getConsentedGoogleVendors(result: FlutterResult) {
    result(ExportData().getConsentedGoogleVendors()?.description)
  }

  private func getConsentedOtherVendors(result: FlutterResult) {
    result(ExportData().getConsentedOtherVendors()?.description)
  }

  private func getConsentedOtherLiVendors(result: FlutterResult) {
    result(ExportData().getConsentedOtherLiVendors()?.description)
  }

  private func getConsentedNonTcfPurposes(result: FlutterResult) {
    result(ExportData().getConsentedNonTcfPurposes()?.description)
  }

  private func getGoogleConsentMode(result: FlutterResult) {
    let googleConsent = ExportData().getGoogleConsentMode()
    let formatted = """
      Analytics Storage: \(googleConsent?.analyticsStorageGranted ?? false),
      Ad Storage: \(googleConsent?.adStorageGranted ?? false),
      Ad User Data: \(googleConsent?.adUserDataGranted ?? false),
      Ad Personalization: \(googleConsent?.adPersonalizationGranted ?? false)
      """

    result(formatted)
  }
}
